### PR TITLE
Refactor Nomad Cluster Setup

### PR DESCRIPTION
**Pull request by Ranjandas**
This is a combination of multiple PRs (#23, #24, #25)

In addition to the PRs, here are two additional changes:
* Install CNI Plugins to support connect-enabled jobs
* Update the Nomad User ACL policy to have `alloc-exec` permission.

---
#23 
The Ubuntu Xenial image is not available anymore officially from Canonical on AWS (except the Pro versions).

The following changes were introduced as part of the AMI change:

* Switch to the latest LTS release (22.04/jammy)
* Switch to using the `ip` command instead of `ifconfig` (deprecated)
* Introduced a system-resolved drop-in for Consul DNS access and stopped Consul from listening on port 53. This is because systemd-resolved already binds to port 53, and Consul won't start.

  The drop-in configuration also listens on the Docker bridge IP so that Nomad tasks can resolve Consul DNS names by overriding DNS by pointing to the Docker bridge IP
* Explicitly added the universe repository without which `jq` and `redis-tools` installation were failing

In addition, `packer {}` block was added so that `packer init` installs the required plugins for the `build` to work.

Fixes the following issues: 
* https://github.com/hashicorp/learn-nomad-cluster-setup/issues/20
* https://github.com/hashicorp/learn-nomad-cluster-setup/issues/15
* https://github.com/hashicorp/learn-nomad-cluster-setup/issues/13

---
#24 
Update the anonymous token policy so that the Consul DNS resolution works out of the box.

---
#25 
* Updated Nomad, Consul, Vault and Consul-Template versions to their latest
* Switched from binary installation to native Linux packages (Apt)
* Removed systemd unit files for those that already have it as part of the package
* Added `zone_pattern` for `retry_join` for faster discovery in GCP

### CODE REVIEW

**Comment by EugenKon** [comment]
Great job, man! Can you review `submit-job` capability? It looks like there should be write policy to be ably to submit, but I am not familiar with Nomad's ACL.
Related: https://github.com/hashicorp/learn-nomad-cluster-setup/issues/28

File: aws/image.pkr.hcl [hunk 1]
```diff
@@ -1,3 +1,13 @@
+packer {
+  required_plugins {
+    amazon = {
+      source  = "github.com/hashicorp/amazon"
+      version = "~> 1"
```
Comment (tunzor) [inline aws/image.pkr.hcl:5]:
```suggestion
      version = ">= 1.3.1"
```

File: azure/image.pkr.hcl [hunk 2]
```diff
@@ -1,3 +1,12 @@
+packer {
+  required_plugins {
+    azure = {
+      source  = "github.com/hashicorp/azure"
+      version = "~> 2"
```
Comment (tunzor) [inline azure/image.pkr.hcl:5]:
```suggestion
      version = ">= 2.0.5"
```

File: gcp/image.pkr.hcl [hunk 3]
```diff
@@ -1,3 +1,12 @@
+packer {
+  required_plugins {
+    googlecompute = {
+      version = ">= 1.1.4"
+      source  = "github.com/hashicorp/googlecompute"
```
Comment (tunzor) [inline gcp/image.pkr.hcl:5]:
```suggestion
      source  = "github.com/hashicorp/googlecompute"
      version = ">= 1.1.4"
```

File: shared/data-scripts/user-data-server.sh [hunk 4]
```diff
@@ -51,6 +51,15 @@ consul acl role create -name "nomad-auto-join" -description "Role with policies
 
 consul acl token create -accessor=${nomad_consul_token_id} -secret=${nomad_consul_token_secret} -description "Nomad server/client auto-join token" -role-name nomad-auto-join -token-file=$CONSUL_BOOTSTRAP_TOKEN
 
+# Update anonymous token policy to support DNS lookup against Consul DNS
+consul acl policy create \
+    -name anonymous-policy \
+    -rules 'service_prefix "" { policy = "read" } node_prefix "" { policy = "read" }' \
+    -token-file $CONSUL_BOOTSTRAP_TOKEN
+
+consul acl token update -id anonymous -policy-name anonymous-policy -token-file $CONSUL_BOOTSTRAP_TOKEN
```
Comment (tunzor) [inline shared/data-scripts/user-data-server.sh:60]: @Ranjandas we're not doing any DNS lookups with consul as part of the cluster setup - is this meant for user's services that may be running on the cluster post-setup? 

Adding the permissions to the anonymous policy exposes the service/node info to anyone with the URL and is not a thing we want to do so I just want to understand what the intention was here.

**Review by tunzor — COMMENTED** [review]
Thanks for the refactor work, it's much appreciated 👍 

Tested the aws scenario and it works great, going to test azure and gcp then we should be good to merge. There are updates to the tutorials as well so I'll coordinate those with the merging of this PR.

File: shared/data-scripts/user-data-server.sh (hunk 4 above)
Comment (EugenKon) [reply shared/data-scripts/user-data-server.sh:60]: I suppose there should be an option to allow DNS queries only from private IPs.

File: gcp/image.pkr.hcl [hunk 5]
```diff
@@ -13,7 +22,7 @@ variable "zone" {
 source "googlecompute" "hashistack" {
   image_name   = "hashistack-${local.timestamp}"
   project_id   = var.project
-  source_image = "ubuntu-minimal-1804-bionic-v20221026"
+  source_image = "ubuntu-minimal-2204-jammy-v20231213b"
```
Comment (tunzor) [inline gcp/image.pkr.hcl:25]:
```suggestion
  source_image = "ubuntu-minimal-2204-jammy-v20240318"
```
Newer base image available

**Review by tunzor — CHANGES_REQUESTED** [review]
Requesting a few changes and clarifications

File: shared/data-scripts/user-data-server.sh (hunk 4 above)
Comment (Ranjandas) [reply shared/data-scripts/user-data-server.sh:60]: Hi @tunzor, @im2nguyen, the service discovery in Nomad is by querying the API using templating right? (As shown in this tutorial https://developer.hashicorp.com/nomad/tutorials/service-discovery/service-discovery-app-deployment), and not by native DNS query. This change is to allow two services talk to each other purely using Consul DNS Names. Please let me know if this makes sense.

Comment (tunzor) [reply shared/data-scripts/user-data-server.sh:60]: Ahh I understand what I've missed now, my mistake for conflating the two. I was testing with jobs that use both Nomad and Consul service discovery.

I'm going to talk through these changes more with the education team and see if we can have Consul DNS accessible without over-privileging the anonymous token and not adding too much more to the tutorials.

**Comment by david-yu** [comment]
@tunzor Could you take over the PR from here and help get this PR merged? Thank you!

**Comment by tunzor** [comment]
@Ranjandas I figured out a way around updating the anonymous policy by adding the agent token to the Consul client configuration but couldn't add the suggestions to this PR so I opened another one that includes your other changes.

Can you double check PR #29 and make sure it's complete? I've tested it and DNS lookups work as they should but feel free to test it out if you like.

**Comment by tunzor** [comment]
Closing this PR as the changes have been merged as part of #29, thank you very much for your contributions!

**Pull request closed without merging**